src/bin/pgcopydb/ld_replay.c

```diff
@@ -249,3 +250,136 @@ stream_replay_line(void *ctx, const char *line, bool *stop)
 
 	return true;
 }
+
+
+/* streamReplayLineIntoBuffer is a callback function for the ReadFromStreamContext
+ * and read_from_stream infrastructure. It's called on each line read from a
+ * stream such as a unix pipe and buffers the line into a file until we encounter
+ * a COMMIT. This ensures that we can replay the whole transaction in one go
+ * and avoids deadlock when the BEGIN messages arrives before the existence of
+ * transation metadata file which is created by the transform process.
+ */
+static bool
+streamReplayLineIntoBuffer(void *ctx, const char *line, bool *stop)
+{
+	ReplayStreamCtx *replayCtx = (ReplayStreamCtx *) ctx;
+	StreamApplyContext *context = &(replayCtx->applyContext);
+
+	if (strstr(line, OUTPUT_BEGIN) == line)
+	{
+		/* If there is already a replay buffer open, it indicates a partial
+		 * transaction. Close it and start a new one.
+		 */
+		if (replayCtx->txnReplayBuffer != NULL)
+		{
+			fclose(replayCtx->txnReplayBuffer);
+		}
```

Comment:
I'm not sure I follow how this is not an ERROR (or a WARN) situation? When does that happen in a normal execution?

Reply:
The idea is to use this for prefetch mode as well, during the resume, we might encouter partially written txns. But that doesn't apply for replay. I would rather add an error message here.

Reply:
Even in prefetch mode we would only encounter a single partial transaction I believe, the last one we received. Cases where we would deal with a partial transaction and then another transaction certainly indicate a bug?

Reply:
>Even in prefetch mode we would only encounter a single partial transaction I believe, the last one we received.

This is what is happening. We have to ignore the partial transactions (i.e. either a COMMIT without a BEGIN or BEGIN without COMMIT).

Reply:
I addressed all the review comments except this.

Reply:
I think I still have some with the code (the comments are good at explaining the situation, thanks). The trouble is that we don't protect against applying this approach even in the middle of our operations, when it looks like it's only needed before we reach our own “consistent point”, never after that. Can we make sure that as soon as we start replaying transactions, then we raise the situation from a DEBUG information to an actual ERROR? (well assuming my analysis is correct here).